support/builders/amlogic-g12/default.nix

```diff
@@ -0,0 +1,74 @@
+{ lib, buildTowBoot, meson64-tools, imageBuilder, runCommandNoCC, spiInstallerPartitionBuilder }:
+
+let self =
+{ defconfig, FIPDIR, withSPI ? false, postBuild ? "",  ...} @ args':
+
+let
+  args = removeAttrs args' (builtins.attrNames (builtins.functionArgs self));
+
+  partitionOffset = 1;
+  sectorSize = 512;
+  firmwareMaxSize = imageBuilder.size.MiB 4;
+
+  firmwarePartition = imageBuilder.firmwarePartition {
+    inherit sectorSize;
+    partitionOffset = partitionOffset; # in sectors
+    partitionSize = firmwareMaxSize; # in bytes
+    firmwareFile = "${firmware}/u-boot.bin";
+  };
+
+  baseImage' = extraPartitions: imageBuilder.diskImage.makeMBR {
+    name = "disk-image";
+    diskID = "01234567";
+    # Align on 512 sector sizes.
+    # This is not for the actual partitions for the end-user, only so we
+    # are able to make the first partition start at sector 0x01
+    alignment = 512;
+
+    partitions = [
+      firmwarePartition
+    ] ++ extraPartitions;
+  };
+  baseImage = baseImage' [];
+  spiInstallerImage = baseImage' [
+    (spiInstallerPartitionBuilder {
+      inherit defconfig;
+      firmware = "${firmware}/u-boot.bin";
+    })
+  ];
+
+  firmware = buildTowBoot ({
+    inherit defconfig FIPDIR;
+
+    nativeBuildInputs = [
+      meson64-tools
+    ];
+
+    postBuild = ''
+      pkg --type bl30 --output bl30.pkg $FIPDIR/bl30.bin $FIPDIR/bl301.bin
+      pkg --type bl2 --output bl2.pkg $FIPDIR/bl2.bin $FIPDIR/acs.bin
+      bl30sig --input bl30.pkg --output bl30.30sig
+      bl3sig  --input bl30.30sig --output bl30.3sig
+      bl3sig  --input $FIPDIR/bl31.img --output bl31.3sig
+      bl3sig  --input u-boot.bin --output bl33.3sig
+      bl2sig  --input bl2.pkg --output bl2.2sig
+      args=()
+      fws=(ddr4_1d.fw ddr4_2d.fw ddr3_1d.fw piei.fw lpddr4_1d.fw lpddr4_2d.fw diag_lpddr4.fw aml_ddr.fw lpddr3_1d.fw)
+      for i in "''${!fws[@]}"; do
+        fw="$FIPDIR/''${fws[i]}"
+        test -e $fw && args+=("--ddrfw$((i+1))" $fw)
+      done
+      bootmk --output u-boot2.bin --bl2 bl2.2sig --bl30 bl30.3sig --bl31 bl31.3sig --bl33 bl33.3sig "''${args[@]}"
+      mv u-boot2.bin u-boot.bin
```

Comment:
I'm pretty sure I'm being rude right here, but please bear with me :)

I would prefer if this segment from the Nixpkgs amlogic PR was used outright:

 - https://github.com/NixOS/nixpkgs/blob/f02c05e056b9bef1b8ebc445a3046432a68d4314/pkgs/misc/uboot/amlogic.nix#L36-L88

After all, I wrote it entirely considering it would be used in Tow-Boot once I had someone handy to test and validate it works.

While it's mostly stylistic differences, there is one huge difference: `ddrfwX` files are used specifically by name and index, rather than *tried*. like you do here. While your approach is probably better in many ways, I'm not able to confirm that it is good. My main concern is that if for a board one of the firmware from 1 through 7 is missing, we want to break. We want to break as we are not expecting that any of them are missing *going on from all the current knowledge*.

Reply:
What board doesn't need ddrfw8?

Reply:
Compare those listings:

 - https://github.com/LibreELEC/amlogic-boot-fip/tree/master/odroid-n2
 - https://github.com/LibreELEC/amlogic-boot-fip/tree/master/odroid-hc4

This is for one vendor.


Then another sample:

 - https://github.com/LibreELEC/amlogic-boot-fip/tree/master/sei510
 - https://github.com/LibreELEC/amlogic-boot-fip/tree/master/sei610

So it's definitely not unexpected to have only those two missing. Furthermore IIRC LibreELEC also has a similar (but differently implemented) logic, where ddrfw8/9 end up being conditional on the files existing. And all fw files are at *specific numbered offsets*